test/testpushnotifications.cpp

```diff
@@ -0,0 +1,351 @@
+#include <QVector>
+#include <QWebSocketServer>
+#include <QSignalSpy>
+
+#include "account.h"
+#include "pushnotifications.h"
+#include "mockwebsocketserver.h"
+#include "creds/abstractcredentials.h"
+
+using namespace std::chrono_literals;
+
+class CredentialsStub : public OCC::AbstractCredentials
+{
+    Q_OBJECT
+
+public:
+    CredentialsStub(const QString &user, const QString &password)
+        : _user(user)
+        , _password(password)
+    {
+    }
+    virtual QString authType() const { return ""; }
+    virtual QString user() const { return _user; }
+    virtual QString password() const
+    {
+        return _password;
+    }
+    virtual QNetworkAccessManager *createQNAM() const { return nullptr; }
+    virtual bool ready() const { return false; }
+    virtual void fetchFromKeychain() { }
+    virtual void askFromUser() { }
+
+    virtual bool stillValid(QNetworkReply * /*reply*/) { return false; }
+    virtual void persist() { }
+    virtual void invalidateToken() { }
+    virtual void forgetSensitiveData() { }
+
+private:
+    QString _user;
+    QString _password;
+};
+
+OCC::AccountPtr createAccount()
+{
+    auto account = OCC::Account::create();
+
+    QStringList typeList;
+    typeList.append("files");
+
+    QString websocketUrl("ws://localhost:12345");
+
+    QVariantMap endpointsMap;
+    endpointsMap["websocket"] = websocketUrl;
+
+    QVariantMap notifyPushMap;
+    notifyPushMap["type"] = typeList;
+    notifyPushMap["endpoints"] = endpointsMap;
+
+    QVariantMap capabilitiesMap;
+    capabilitiesMap["notify_push"] = notifyPushMap;
+
+    account->setCapabilities(capabilitiesMap);
+
+    return account;
+}
+
+std::unique_ptr<OCC::PushNotifications> createPushNotifications(OCC::Account *account)
+{
+    auto pushNotifications = std::make_unique<OCC::PushNotifications>(account);
+    pushNotifications->setReconnectTimerInterval(0);
+
+    return std::move(pushNotifications);
+}
+
+class TestPushNotifications : public QObject
+{
+    Q_OBJECT
+
+private slots:
+    void testSetup_correctCredentials_authenticatedOnWebSocket()
+    {
+        const QString user = "user";
+        const QString password = "password";
+        MockWebSocketServer mockServer;
+        QSignalSpy spy(&mockServer, &MockWebSocketServer::processTextMessage);
+        QVERIFY(spy.isValid());
+        auto account = createAccount();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account.data());
+
+        // Setup push notifications
+        pushNotifications->setup();
+
+        // Wait for authentication
+        spy.wait();
+
+        // Right authentication data should be sent
+        QCOMPARE(spy.count(), 2);
+
+        const auto userSent = spy.at(0).at(1).toString();
+        const auto passwordSent = spy.at(1).at(1).toString();
+
+        QCOMPARE(userSent, user);
+        QCOMPARE(passwordSent, password);
+    }
```

Comment:
Note that it'd feel more natural to me to test all the way to the ready() signal here. Because you'd be testing the authenticate scenario in full. Separating this in two tests gives an after taste of testing too much against how the implementation is structured.

Would also lead to less code duplication between tests since testAuthentication_correctCredentials_emitReady would be gone.